File: src/usage.ts
/* What the agents have cost, sliced the ways the question gets asked.
 *
 * The buckets come from `usage.rs` already grouped by day, agent, model and
 * project — one row per combination per day rather than one per run — so
 * everything here is a fold over a list that is already small.
 *
 * ⚠️ Pure, and tested without a DOM, for the same reason `spend.ts` is: the
 * arithmetic IS the feature. A total that is merely plausible still renders,
 * and nothing about the panel would look wrong if it were double-counting.
 */

export interface Bucket {
  day: string;
  provider: string;
  /** As the provider names it. Empty where the transcript never said. */
  model: string;
  project: string;
  input: number;
  output: number;
  runs: number;
  seconds: number;
}

/** One row of a breakdown, whatever it was cut by. */
export interface Slice {
  /** What it was cut by — a provider, a model id, a project name. */
  key: string;
  input: number;
  output: number;
  runs: number;
  seconds: number;
  /** For an agent row: the model it spent the most on. ⚠️ The *biggest*, not
   *  the newest — a row that named whichever model answered last would change
   *  under you for a single cheap question. */
  top?: string;
}

export function spent(row: { input: number; output: number }): number {
  return row.input + row.output;
}

function fold(buckets: Bucket[], key: (bucket: Bucket) => string): Map<string, Slice> {
  const out = new Map<string, Slice>();
  for (const bucket of buckets) {
    const id = key(bucket);
    const row = out.get(id) ?? { key: id, input: 0, output: 0, runs: 0, seconds: 0 };
    row.input += bucket.input;
    row.output += bucket.output;
    row.runs += bucket.runs;
    row.seconds += bucket.seconds;
    out.set(id, row);
  }
  return out;
}

const bigger = (a: Slice, b: Slice) => spent(b) - spent(a);

/** Per agent, biggest spender first, each naming the model it spent most on. */
export function byAgent(buckets: Bucket[]): Slice[] {
  const rows = fold(buckets, bucket => bucket.provider);
  for (const [provider, row] of rows) {
    const models = fold(
      buckets.filter(bucket => bucket.provider === provider && bucket.model),
      bucket => bucket.model,
    );
    row.top = [...models.values()].sort(bigger)[0]?.key;
  }
  return [...rows.values()].filter(row => spent(row) > 0).sort(bigger);
}

/** Per model, biggest first. ⚠️ Buckets with no model are dropped rather than
 *  gathered under a blank row: "" is not a model, it is a session this app
 *  started watching mid-run, and a nameless row at the top of a list about
 *  which model costs what says nothing anybody can act on. */
export function byModel(buckets: Bucket[]): Slice[] {
  const rows = fold(buckets.filter(bucket => bucket.model), bucket => bucket.model);
  return [...rows.values()].filter(row => spent(row) > 0).sort(bigger);
}

/** Per project, biggest first. */
export function byProject(buckets: Bucket[]): Slice[] {
  const rows = fold(buckets, bucket => bucket.project);
  return [...rows.values()].filter(row => spent(row) > 0).sort(bigger);
}

/** `2026-09-18` minus n days, in the same format. */
export function dayBefore(day: string, back: number): string {
  const [year, month, date] = day.split("-").map(Number);
  /* ⚠️ UTC, deliberately, though these are LOCAL days. The arithmetic is
   * "subtract 24 hours n times", and in local time that is wrong twice a year:
   * the day the clocks go back has 25 hours in it, so a local-midnight Date
   * walks onto the same calendar day twice and the chart grows a duplicate
   * column. The strings are already local days; only the counting is UTC. */
  const at = Date.UTC(year, month - 1, date) - back * 86_400_000;
  return new Date(at).toISOString().slice(0, 10);
}

/** The last `days` days ending today, oldest first, zeros included.
 *
 * ⚠️ Every day, not only the ones with a bucket. A chart built from the days
 * that have data has no gaps in it — so a week with three days off looks like
 * three days of work in a row, which is the opposite of what it says.
 */
export function byDay(buckets: Bucket[], today: string, days: number): Slice[] {
  const rows = fold(buckets, bucket => bucket.day);
  const out: Slice[] = [];
  for (let back = days - 1; back >= 0; back--) {
    const day = dayBefore(today, back);
    out.push(rows.get(day) ?? { key: day, input: 0, output: 0, runs: 0, seconds: 0 });
  }
  return out;
}

/** The whole lot, however it was cut. */
export function all(rows: Slice[]): Slice {
  return rows.reduce(
    (sum, row) => ({
      key: "",
      input: sum.input + row.input,
      output: sum.output + row.output,
      runs: sum.runs + row.runs,
      seconds: sum.seconds + row.seconds,
    }),
    { key: "", input: 0, output: 0, runs: 0, seconds: 0 },
  );
}

File: src-tauri/src/usage.rs
//! What the agents have cost, kept.
//!
//! `runlog.rs` keeps every finished run for a fortnight, which is what the
//! Review screen reads and is the right shape for "what happened on Tuesday".
//! It is the wrong shape for "is this month worse than last": a run is a few
//! hundred bytes, there are a couple of dozen a day, and two weeks is where
//! that stops being free.
//!
//! So this is the other half — one bucket per day per agent per model per
//! project, merged as runs land. A day of hard work is four or five buckets
//! rather than forty runs, which is small enough to keep for months and is
//! already grouped the way every question about it is asked.
//!
//! ⚠️ **Additive, and never derived from `runs.json`.** Rebuilding these from
//! the runs would mean the totals silently changed shape the day the fortnight
//! rolled over and the oldest runs were dropped — the numbers would still
//! render, and they would be wrong about exactly the period this file exists
//! to remember.

use chrono::Local;
use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Manager};

const FILE: &str = "usage.json";

/// Four months. Long enough for "is this month worse than last" to have an
/// answer, short enough that the file stays tens of kilobytes.
const RETAIN_DAYS: i64 = 120;

/// One day of one model, on one project.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Bucket {
    /// Local day, `YYYY-MM-DD`. ⚠️ Stored rather than derived from a
    /// timestamp, so a run stays filed under the day it *felt* like whatever
    /// the reader's timezone does later.
    pub day: String,
    pub provider: String,
    /// The model, as the provider names it. Empty where the transcript never
    /// said — which is normal for a session that was picked up mid-run.
    #[serde(default)]
    pub model: String,
    pub project: String,
    pub input: u64,
    pub output: u64,
    pub runs: u32,
    pub seconds: u64,
}

impl Bucket {
    /// Whether two runs belong in the same bucket.
    fn same(&self, other: &Bucket) -> bool {
        self.day == other.day
            && self.provider == other.provider
            && self.model == other.model
            && self.project == other.project
    }
}

#[derive(Default)]
pub struct Store(pub std::sync::Mutex<Vec<Bucket>>);

fn cutoff() -> String {
    (Local::now() - chrono::Duration::days(RETAIN_DAYS))
        .format("%Y-%m-%d")
        .to_string()
}

pub fn load(app: &AppHandle) {
    let mut stored: Vec<Bucket> = crate::config::load_beside(FILE).unwrap_or_default();
    let floor = cutoff();
    stored.retain(|bucket| bucket.day >= floor);
    /* ⚠️ Seeded from the runs ONCE, when this file does not exist yet.
     *
     * Deriving these from `runs.json` in general is wrong — it drops its own
     * history after a fortnight, so the totals would silently change shape the
     * day the window rolled over. But on the very first run there is nothing
     * to be wrong about and a fortnight of real history sitting one file away:
     * without this the chart is empty for two weeks, which reads as a feature
     * that does not work rather than as one that has not been fed yet.
     *
     * ⚠️ Runs from before providers were recorded default to Claude, which is
     * what they were: it is the only agent this app could see. */
    if stored.is_empty() {
        if let Ok(runs) = app.state::<crate::runlog::Store>().0.lock() {
            for run in runs.iter() {
                let fresh = Bucket {
                    day: run.day.clone(),
                    provider: run.provider.clone(),
                    model: run.model.clone().unwrap_or_default(),
                    project: run.project.clone(),
                    input: run.input,
                    output: run.output,
                    runs: 1,
                    seconds: run.seconds,
                };
                match stored.iter_mut().find(|bucket| bucket.same(&fresh)) {
                    Some(bucket) => {
                        bucket.input += fresh.input;
                        bucket.output += fresh.output;
                        bucket.runs += 1;
                        bucket.seconds += fresh.seconds;
                    }
                    None => stored.push(fresh),
                }
            }
        }
        if !stored.is_empty() {
            crate::config::save_beside(FILE, &stored);
        }
    }
    if let Ok(mut held) = app.state::<Store>().0.lock() {
        *held = stored;
    }
}

/// Fold one finished run into its day.
pub fn record(app: &AppHandle, run: &crate::sessions::Finished) {
    let fresh = Bucket {
        day: crate::runlog::today(),
        provider: run.provider.clone(),
        model: run.model.clone().unwrap_or_default(),
        project: run.project.clone(),
        input: run.input,
        output: run.output,
        runs: 1,
        seconds: run.seconds,
    };
    let snapshot = {
        let state = app.state::<Store>();
        let Ok(mut held) = state.0.lock() else { return };
        match held.iter_mut().find(|bucket| bucket.same(&fresh)) {
            Some(bucket) => {
                bucket.input += fresh.input;
                bucket.output += fresh.output;
                bucket.runs += 1;
                bucket.seconds += fresh.seconds;
            }
            None => held.push(fresh),
        }
        let floor = cutoff();
        held.retain(|bucket| bucket.day >= floor);
        held.clone()
    };
    crate::config::save_beside(FILE, &snapshot);
}

/// Every bucket from the last `days` days, oldest first.
///
/// ⚠️ `days` counts back from today INCLUSIVE, so 1 is today and 7 is the week
/// ending now. Off by one here is a chart that silently starts on Tuesday.
#[tauri::command]
pub fn get_usage(app: AppHandle, days: i64) -> Vec<Bucket> {
    let floor = (Local::now() - chrono::Duration::days(days.max(1) - 1))
        .format("%Y-%m-%d")
        .to_string();
    let state = app.state::<Store>();
    let Ok(held) = state.0.lock() else {
        return Vec::new();
    };
    let mut out: Vec<Bucket> = held.iter().filter(|b| b.day >= floor).cloned().collect();
    out.sort_by(|a, b| a.day.cmp(&b.day));
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    fn bucket(day: &str, provider: &str, model: &str, project: &str) -> Bucket {
        Bucket {
            day: day.into(),
            provider: provider.into(),
            model: model.into(),
            project: project.into(),
            input: 10,
            output: 1,
            runs: 1,
            seconds: 5,
        }
    }

    #[test]
    fn a_bucket_is_one_day_one_model_one_project() {
        let a = bucket("2026-09-18", "claude", "claude-opus-5", "akcesfonia");
        assert!(a.same(&bucket("2026-09-18", "claude", "claude-opus-5", "akcesfonia")));
        // ⚠️ The model is part of the key. Folded together, "what did Opus
        // cost me" has no answer — which is most of the reason to keep this.
        assert!(!a.same(&bucket("2026-09-18", "claude", "claude-haiku-4-5", "akcesfonia")));
        assert!(!a.same(&bucket("2026-09-18", "codex", "claude-opus-5", "akcesfonia")));
        assert!(!a.same(&bucket("2026-09-17", "claude", "claude-opus-5", "akcesfonia")));
        assert!(!a.same(&bucket("2026-09-18", "claude", "claude-opus-5", "esono")));
    }

    #[test]
    fn the_window_is_kept_by_string_comparison() {
        // The whole retention rule, and it works only because the format is
        // zero-padded and big-endian. Same rule as `runlog`.
        let floor = cutoff();
        assert!(floor < crate::runlog::today());
        assert!("2026-01-02" < "2026-01-10");
        assert!("2025-12-31" < "2026-01-01");
    }
}
